plutus-core/untyped-plutus-core/src/Data/BRAL.hs

```diff
@@ -0,0 +1,108 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE LambdaCase   #-}
+module Data.BRAL ( BList -- abstract type for now
+                 , nil
+                 , cons
+                 , index
+                 , indexZero
+                 , Data.BRAL.null
+                 , Data.BRAL.head
+                 , Data.BRAL.tail
+                 , uncons
+                 ) where
+
+import           Data.Bits  (unsafeShiftR)
+import           Data.Maybe (fromMaybe)
+
+-- A complete binary tree.
+-- Note: the size of the tree is not stored/cached,
+-- unless it appears as a root tree in BList, which the size is stored inside the Cons.
+data Tree a = Leaf a
+            | Node a !(Tree a) !(Tree a)
+            deriving (Eq, Show)
+
+-- a strict list of complete binary trees accompanied by their size.
+-- The trees appear in >=-size order.
+-- Note: this blist is strict on its spine, unlike the haskell's stdlib list
+data BList a = Cons
+               -- TODO: use arch-independent Word32 or Word64
+               {-# UNPACK #-} !Word -- ^ the size of the head tree
+               !(Tree a) -- ^ the head tree
+               !(BList a) -- ^ the tail trees
+             | Nil
+             deriving (Eq, Show)
```

Comment:
I think an `IsList` instance is a good idea, will make  your tests easier.

Reply:
Probably worth adding a bunch of "standard" instances tbh.

Reply:
Probably worth a comment about `Eq`, actually: I thought "hang on, shouldn't we be checking the contents are the same?", but this works since the BRAL representation is unique for a given number of elements :+1: